File: 0426-convert-binary-search-tree-to-sorted-doubly-linked-list/0426-convert-binary-search-tree-to-sorted-doubly-linked-list.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val,Node _left,Node _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};
*/

class Solution {
    Node head, tail;
    public Node treeToDoublyList(Node root) {
        if (root == null)
            return root;
        
        head = null;
        tail = null;
        
        inOrderLinking(root);
        
        head.left = tail;
        tail.right = head;
        
        return head;
    }
    
    private void inOrderLinking(Node root) {
        // go till left most node
        if (root != null) {
            inOrderLinking(root.left);
            
            // now we're sure there no left most node
            if (tail == null) {
                head = root;
            // if there is tail then append new ele to end
            } else {
                root.left = tail;
                tail.right = root;
            }
                
            tail = root;
            
            inOrderLinking(root.right);
        }
    }
}

File: 0426-convert-binary-search-tree-to-sorted-doubly-linked-list/0426-convert-binary-search-tree-to-sorted-doubly-linked-list.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
"""

class Solution:
    def treeToDoublyList(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return root
        
        self.head = None
        self.tail = None
        
        self.inOrderLinking(root)
        
        self.head.left = self.tail
        self.tail.right = self.head
        
        return self.head
    
    def inOrderLinking(self, root):
        # go till left most node
        if root:
            self.inOrderLinking(root.left)
            
            # now we're sure there no left most node
            if not self.tail:
                self.head = root
            # if there is tail then append new ele to end
            else:
                root.left = self.tail
                self.tail.right = root
                
            self.tail = root
            
            self.inOrderLinking(root.right)